test.py async and schema changes

alecco opened the pull request:
Change tests to use async mode and add helpers and tests for schema changes.

These test series will be expanded with topology changes.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/312/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/340/

alecco commented:
@nyh could you be so kind to have a look?

(in spite of Kostja's not yet queued and CI failure)

I'll later create separate branches for rebases and fixes so the incremental changes can be easily compared.

cheers

nyh commented on test/pylib/schema_helper.py:
```diff
@@ -0,0 +1,238 @@
+        await asyncio.gather(*(t.drop() for t in self.tables))
+        self.removed_tables.extend(self.tables)
+
+    async def verify_schema(self, table: Union[Table, str] = None) -> None:
```
What's the purpose of this function?
If this is meant to *test* something, i.e., that the system_schema keyspace contains the right things, then this code should be in a test - not in a utility class - and this test should explain what it's testing and why.
If it's really a utility function, and not a test, I'm afraid I don't understand it's utility.

nyh replied:
I'm giving up. I'm personally not thrilled when complex highly-specific code that is designed for one test only is put in a library that is supposedly useful for many tests, but maybe I'm missing the bigger picture and you intend this to be used in additional tests as well?

In any case, I won't insist, but please do answer. Thanks.

alecco replied:
We need to test schema changes on top of topology changes when using Raft. The whole point of this file is to provide helpers to these upcoming tests so we don't need to manually write schema checks with CQL and tracking state of the table. Docstrings were added in v7.

nyh commented on test/pylib/schema_helper.py:
```diff
@@ -0,0 +1,238 @@
+
+async def create_keyspace(cql: CassandraSession):
+    keyspace_name = f"k_{new_keyspace_id():02}"
+    dc = await this_dc(cql)
```
nitpick: It's a waste to find this_dc(cql) on every create_keyspace from scratch.
It's not important because there is no reason to create many keyspaces in a test anyway.

alecco replied:
Removed in v3

nyh commented on test/topology/test_null.py:
```diff
@@ -5,23 +5,18 @@
     with pytest.raises(InvalidRequest, match='Key may not be empty'):
-        await cql.run_async(f"DELETE FROM {table1} WHERE p='' AND c='{s}'")
+        await cql.run_async(f"DELETE FROM {tables[0]} WHERE pk = '' AND {tables[0].columns[1].name} = '{s}'")
+    await tables.verify_schema()
```
Again, I don't understand what this tables.verify_schema() is supposed to be doing or testing in a test called test_delete_empty_string_key.
Also where does the test table get deleted here?

nyh replied:
I'm still waiting to understand why test_null.py and its test_delete_empty_string_key even exists. Why not remove it and start with a real test?

alecco replied:
test_null was removed in v8

nyh commented on test/topology/test_null.py:
```diff
@@ -5,23 +5,18 @@
     s = "foobar"
     # An empty-string clustering *is* allowed:
-    await cql.run_async(f"DELETE FROM {table1} WHERE p='{s}' AND c=''")
+    await cql.run_async(f"DELETE FROM {tables[0]} WHERE pk = '{s}' AND {tables[0].columns[1].name} = ''")
```
One of the things I don't like about large utilities like that "schema helper" is that it obscures what this test is really testing:
1. Does the table returned by get_schema always have a string partition key? Why is this the case, and how does the reader of this test code guess this?
2. Why does this test have 5 columns, and what's the point of them being random?

alecco replied:
test_null removed in v8

PK made random type in v8

avikivity commented on install-dependencies.sh:
```diff
@@ -76,6 +76,7 @@ fedora_packages=(
     python3-tabulate
     python3-boto3
     python3-pytest
+    python3-pytest-asyncio
```
It's better to change it separately, and ahead of time.

alecco commented:
@nyh 

Here are the changes requested without rebasing to Kostja's latest test.py (for convenience and to keep this pull moving forward)

https://github.com/alecco/scylla/compare/raft-test-schema-08-v2..alecco:raft-test-schema-08-10-v3-no-rebase?expand=1

- separate pull req for python3-pytest-asyncio dep   @nyh
- drop session scope commit                          @nyh
- asyncio wrapper back to proper commit              @nyh
- drop all tables after test finishes (use fixture)  @nyh
- removed global conf from test/conftest.py          @nyh
- removed leftover ntables mark from v1              @nyh
- reference issue 1207 for Cassandra issue 10250

Thanks

alecco commented:
Last force push was rebased on merged test.py.v5 by Kostja

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/640/

alecco commented:
Rebased on today's master

alecco commented on test/topology/suite.yaml:
```diff
@@ -5,3 +5,4 @@ topology:
     replication_factor: 3
 disable:
     - test_null
+    - test_cassandra_10250
```
For the Cassandra repro, I added it here because using pytest-async and xfail gives an error:

    RuntimeError: cannot schedule new futures after shutdown

This will all go away once we enable raft in a subsequent pull request and the test is not expected to fail anymore.

nyh replied:
This "cannot schedule new futures after shutdown" thing looks like a bug in the test framework. A bug which you introduced in this series which introduced the futures thing, right? So let's fix it instead of hiding it. Your framework will need to be able to handle xfailing tests as well. Maybe in this case you can say let's wait a few hours or days and it will stop being xfail - but in other cases you may want to have an xfail for a long time and it should be ok.

alecco replied:
test_null was removed in v8 and we haven't seen this happen anymore.

alecco commented on test/conftest.py:
```diff
@@ -134,14 +135,14 @@ def cql_test_connection(cql, request):
-    cql.execute("CREATE KEYSPACE " + name + " WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', '" +
-                this_dc + "' : 1 }")
+    await cql.run_async(f"CREATE KEYSPACE {name} WITH REPLICATION = {{ 'class' : 'NetworkTopologyStrategy', " +
+                        f"'{this_dc}' : '{3}' }}")
```
I addressed this by picking RF from configuration (scylla.yaml). All topology tests are run against the same cluster and RF configuration.

alecco commented:
@nyh 

Here are v4 changes without rebasing on latest master:
https://github.com/alecco/scylla/compare/alecco:raft-test-schema-08-10-v3-no-rebase..raft-test-schema-08-v4-no-rebase?expand=1

V4 changes:

* Drop TestTables keyspace
* revert fixture name keyspace to test_keyspace (needed somewhere else)
* take DC RF from configuration and pass it to test  @nyh 
* split commits for async and DC RF
* fix various comments  @nyh 
* yield instead of return on tables fixture @nyh 
* schema helper fixture scope function @nyh 
* check pre-existing row is returned with new schema after drop column @nyh 
* use fails_without_raft fixture for schema linearization repro   @nyh 
* rebased on latest master

alecco commented:
v5:

- use `pytest_asyncio.fixture`                      @kostja 
- shrink schema serialization repro wait to minimum 1s  @kostja 

While there, importing `pytest_asyncio` solves the problem of tests not run due to missing module (only giving a warning in the log).

Comp with v4  https://github.com/alecco/scylla/compare/raft-test-schema-08-v4-rebased..alecco:raft-test-schema-08-v5?expand=1

nyh commented:
> > Rebased on today's master
> 
> Please edit the top comment (which says things about basing on next, v2 - all of it should probably be removed from the top comment which will become the merge commit message).

@alecco you still haven't done that. All the stuff you see on the top comment will end up the merge-commit message.

nyh commented on test/conftest.py:
```diff
@@ -31,6 +33,8 @@ def pytest_addoption(parser):
     parser.addoption('--ssl', action='store_true',
                      help='Connect to CQL via an encrypted TLSv1.2 connection')
+    parser.addoption('--dc_rf', action='store', default='1',
+                     help='Default datacenter replication factor')
```
Ok. I'm not thrilled with this approach but can accept it. I would have liked to think that we can figure out how many nodes exist in the cluster from within the test, instead of it being passed from the test.py configuration. Making this a configuration option can allow users to (accidentally) create non-sensical setups, like just one node and RF=3.

alecco replied:
removed in v7

nyh commented on test/pylib/schema_helper.py:
```diff
@@ -168,6 +171,27 @@ async def insert_seq(self) -> asyncio.Future:
 
+    async def add_index(self, column: Union[Column, str], name: str = None) -> str:
+        if type(column) is int:
+            assert column >= self.pks, f"Idex on {'PK' if column == 0 else 'CK'} column"
```
My comment is still relevant

alecco replied:
Fixed the assert and its message in v8.

nyh reviewed (approved):
There are a few of my comments still open, I would appreciate if you can answer them and/or fix things if you find it appropriate - but in any case I think the remaining problems are minor so I am marking "Approve". The CI run is still in progress so we can't commit yet anway.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/778/

alecco commented:
(removed from top comment)

It is based on next with [#10381 v3 test.py: add cluster tests](https://github.com/scylladb/scylla/pull/10381)
(but now dequeued)

v2:
- fix test description                                 @nyh 
- add/fix comments                                     @nyh 
- make clear what is expected if bug 10250 happens   @nyh 
- remove ntables marker and set up tables manually     @nyh 
- (asyncio auto mode not used as it's not reliable)    @nyh 
- use xfail for failing cassandra 10250 repro          @nyh 
- move helpers to test/pylib                           @kostja
- string conversion for helper classes
- combine commits for ease of review
- rebase

Tests: (dev, release)

v3:

- separate pull req for python3-pytest-asyncio dep   @nyh
- drop session scope commit                          @nyh
- asyncio wrapper back to proper commit              @nyh
- drop all tables after test finishes (use fixture)  @nyh
- removed global conf from test/conftest.py          @nyh
- removed leftover ntables mark from v1              @nyh
- rebased on latest master

Tests: (dev, release)

nyh commented:
With the latest version all cql-pytest tests failed in the CI because:

```
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --dc_rf=1
  inifile: /jenkins/workspace/releng/Scylla-CI/scylla/test/cql-pytest/pytest.ini
  rootdir: /jenkins/workspace/releng/Scylla-CI/scylla/test/cql-pytest
```

pytest can't just pass this "--dc_rf=1" option to every test. It should only pass it to test suites which want it.
Or not use this option at all (I suggested in a comment above that maybe this option isn't necessary at all, and the test can figure out the number of nodes using code, instead of expecting another option).

alecco commented:
> > > Rebased on today's master
> > 
> > 
> > Please edit the top comment (which says things about basing on next, v2 - all of it should probably be removed from the top comment which will become the merge commit message).
> 
> @alecco you still haven't done that. All the stuff you see on the top comment will end up the merge-commit message.

Sorry about that. done.

alecco commented:
> I'm giving up. I'm personally not thrilled when complex highly-specific code that is designed for one test only is put in a library that is supposedly useful for many tests, but maybe I'm missing the bigger picture and you intend this to be used in additional tests as well?
> 
> In any case, I won't insist, but please do answer. Thanks.

@nyh Apologies, I thought it was clear this was for tests coming in a later Raft schema & topology testing series. I'll update the commit message.

alecco commented:
v6:
- fix top comment of pull req                      @nyh 
- remove dc_rf command option (to be added to API in later series)  @nyh / @kostja 
- reduce 20 to 1 size and time of concurrent schema repro @kostja 
- use pytest_asyncio.fixture instead of pytest.fixture @kostja 
- fix test license @kostja 
- remove unused import Table in concurrent schema repro test
- remove unused fixture cql in concurrent schema repro
- remove spurious list comprehension in concurrent schema repro
- pytest_asyncio import ignore type check in conftest.py

Diff with v5 https://github.com/alecco/scylla/compare/raft-test-schema-08-v5-rebased-20220612..alecco:raft-test-schema-08-v6?expand=1

alecco commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/824/

Failures are caused by old `pytest_asyncio`. @benipeled is looking to add this module from pip instead of system to get a recent version.

benipeled commented:
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/824/
> 
> Failures are caused by old `pytest_asyncio`. @benipeled is looking to add this module from pip instead of system to get a recent version.

Since it's a WARN only, let's leave it - for now - to be managed by dnf and avoid mixing the dnf and pip packages,

Regardless of this case, IMO we should move all python base packages to be managed by pip which usually (as in this case) contains updated versions

alecco commented:
I sent v6 without `pytest_asyncio.fixture`. Beni will take care of this if/when CI fails because of this in the future.

alecco commented:
@nyh about test_null... it's not my code and it was merged before me. I'll discuss with Kostja and will remove or fix in a later series.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/845/

alecco commented:
v6-c
- added missed await

alecco commented on test/topology/test_null.py:
```diff
@@ -8,7 +8,7 @@
 
 
-@pytest.fixture(scope="module")
+@pytest.fixture(scope="session")
```
This was removed in a later version and will try to remove test_null altogether in a subsequent pull.

alecco replied:
This is removed in a later commit anyway.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/960/

alecco commented on test/conftest.py:
```diff
@@ -103,14 +139,14 @@ def cql_test_connection(cql, request):
+# at the end of the session. All tests will reuse the same keyspace.
 @pytest.fixture(scope="session")
-def test_keyspace(cql, this_dc):
+async def keyspace(request, cql, this_dc) -> AsyncGenerator:
```
Renamed back to keyspace.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/964/

alecco commented on test/topology/test_schema.py:
```diff
@@ -0,0 +1,49 @@
+
+
+@pytest.mark.asyncio
+async def test_new_table(cql):
```
Consider it a hello world test so other people can base their tests on this one. I was asked to send a series of schema tests before doing schema on top of topology tests.

alecco commented on test/topology/test_cassandra_10250.py:
```diff
@@ -0,0 +1,67 @@
+
+# https://issues.apache.org/jira/browse/CASSANDRA-10250
+# - Create 20 tables to alter and 20 tables to index
+# - In parallel 20 * create table, and drop/add column and index of previous tables
```
Issue was created in a previous version. Test file was renamed. Comments were added.

alecco commented on test/topology/test_schema.py:
```diff
@@ -47,3 +47,15 @@ async def test_verify_schema(cql):
+
+
+@pytest.mark.asyncio
+@pytest.mark.ntables(1)
```
After refactor this is passed to the add_tables() method.

alecco commented on test/conftest.py:
```diff
@@ -146,3 +147,14 @@ async def test_keyspace(request, cql, this_dc) -> AsyncGenerator:
+@pytest.fixture()
+async def tables(request, cql, test_keyspace) -> AsyncGenerator:
+    """Creates a specified table schema in a given keyspace and 
+       destroys it after finished"""
```
This was renamed, comment improved and docstring removed in v8.

alecco commented on test/conftest.py:
```diff
@@ -146,3 +147,14 @@ async def test_keyspace(request, cql, this_dc) -> AsyncGenerator:
+    """Creates a specified table schema in a given keyspace and 
+       destroys it after finished"""
+    tables = TestTables(request.node.name, cql, test_keyspace)
+    return tables
```
Fixed in  v4

alecco commented on test/conftest.py:
```diff
@@ -146,3 +147,14 @@ async def test_keyspace(request, cql, this_dc) -> AsyncGenerator:
+
+# "schema" fixture: Creates and returns a temporary TestTables object
+# used in tests to make schema changes. Tables are dropped after finished.
+@pytest.fixture()
```
Done in v4

alecco commented on test/conftest.py:
```diff
@@ -146,3 +147,14 @@ async def test_keyspace(request, cql, this_dc) -> AsyncGenerator:
     await cql.run_async("DROP KEYSPACE " + name)
+
+
+# "schema" fixture: Creates and returns a temporary TestTables object
```
Fixed in v4

alecco commented on test/topology/test_schema.py:
```diff
@@ -57,3 +57,12 @@ async def test_new_table_insert_one(cql, tables):
+    table = await tables.add_table(ncolumns=5)
+    await table.insert_seq()
+    await table.drop_column()
+    await tables.verify_schema(table)
```
Done in v4

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/968/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/969/

alecco commented on test/topology/test_cassandra_10250.py:
```diff
@@ -0,0 +1,70 @@
+# - Create 20+20 tables to alter and 20 tables to index
+# - In parallel 20 * create table, and drop/add column and index of previous tables
+@pytest.mark.asyncio
+@pytest.mark.xfail    # TODO: enable again once raft is enabled (also see suite.yaml)
```
Done in v7

alecco commented:
v7 changes:
- rename schema_helper to random_tables and adjust accordingly class and fixture names   @kostja 
- revert pytest_asyncio.fixtures as CI uses older package, Beni will fix again when CI switches to pip packages
- improve docstrings for random_tables modules  @kostja 
- improve typing on random_tables
- improve lint score (mostly indentation)
- add logging of CQL queries on random_tables
- remove now obsolete this_dc fixture within random_tables (as keyspace is provided from previous change)

Compared with v6  https://github.com/alecco/scylla/compare/raft-test-schema-08-v6-rebased..alecco:raft-test-schema-08-v7?expand=1

alecco commented:
v8 changes:
- add again change renaming fixture test_keyspace to keyspace (and in its own commit) @kostja 
- remove redundant docstring to random_tables fixture @nyh 
- remove leftover redundant comment in random_tables module (now docstring)
- improve typing in in random_tables module
- pk with random type in random_tables module (and fixed tests accordingly) @nyh 
- removed test_null  @nyh  @kostja 

Compared with v7 https://github.com/alecco/scylla/compare/raft-test-schema-08-v7..alecco:raft-test-schema-08-v8?expand=1

alecco commented:
changes in v9:
- fix Cassandra 10250 repro to match new random_tables
- rename file to test_concurrent_schema as other related tests might be added in the future here
- fix comment for keyspace fixture

Compared to v8 https://github.com/alecco/scylla/compare/raft-test-schema-08-v8..alecco:raft-test-schema-08-v9?expand=1

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/971/

alecco commented on test/topology/test_cassandra_10250.py:
```diff
@@ -0,0 +1,73 @@
+
+import asyncio
+import pytest
+from pylib.schema_helper import Table, Column, UUIDType, IntType         # type: ignore
```
Solved in v6

alecco commented on test/topology/test_schema.py:
```diff
@@ -0,0 +1,86 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
```
Done in v7

alecco commented:
v10 changes:
- re-submitted as previous CI failed on unrelated Enterprise test
- rebased on latest master, just in case
- while there improved comment on event_loop fix for async @kostja 

v10 vs v9
https://github.com/alecco/scylla/compare/raft-test-schema-08-v9-rebased..alecco:raft-test-schema-08-v10?expand=1

alecco commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/988/

All failures are on completely unrelated tests in debug mode. bptree_stress_test, lwt_test, memtable_test, test_secondary_index, test_storage_service.

kbr- commented:
@nyh could you please re-review?

nyh commented:
The CI keeps failing, this time on something I don't understand... I'll restart it and see what happens...

kbr- commented:
It keeps failing due to an existing problem (the topology suite was disabled because of it); I fixed it here: https://github.com/scylladb/scylla/pull/10861
But otherwise the PR can be reviewed.

nyh commented on test/conftest.py:
```diff
@@ -135,12 +135,12 @@ def cql_test_connection(cql, request):
 # tests will reuse the same keyspace.
 @pytest.fixture(scope="session")
-def test_keyspace(cql, this_dc):
+def keyspace(cql, this_dc):
```
I can't comment on the commit message so I'll do it here:
I'm scared of the comment on a clash with a fixture of cql-pytest, and also something that someone else said yesterday on the mailing list. @kostja "promised" me that the top-level test/conftest.py, which I still think was a mistake, only takes effect for tests that "opt in" to use it, and cql-pytest, alternator, redis, scylla-gdb, and others, will NOT use it. Now it appears that they do. This should be fixed instead of working around the problems it causes.

alecco replied:
There is no actual name clash, I got mistaken. I updated the commit message.

nyh commented:
> It keeps failing due to an existing problem (the topology suite was disabled because of it); I fixed it here: #10861 But otherwise the PR can be reviewed.

I reviewed it and approved it for a second time. I'll commit it when the CI passes. If a different fix needs to be committed first, let's do it. I'll take a look.

nyh commented:
I think @kbr- 's change to suite.yaml in https://github.com/scylladb/scylla/pull/10861 will conflict with a change in this series?

kbr- commented:
Rebase shouldn't be a problem.

alecco commented on test/conftest.py:
```diff
@@ -137,6 +137,29 @@ def cql_test_connection(cql, request):
 
+# While the raft-based schema modifications are still experimental and only
+# optionally enabled (the "--raft" option to test/cql-pytest/run enables it
+# in Scylla), some tests are expected to fail on Scylla without this option
```
Fixed in latest

alecco commented:
> I think @kbr- 's change to suite.yaml in #10861 will conflict with a change in this series?

I'll send now a version rebased on top of that branch.

alecco commented:
v12 changes:
- rebased on pending pull #10861          @nyh @kbr- 
- fixed comment of check_pre_raft fixture  @nyh 
- fixed commit message of keyspace fixture  @nyh 

https://github.com/alecco/scylla/compare/raft-test-schema-08-v10-rebased..alecco:raft-test-schema-08-v12?expand=1

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1016/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1019/

alecco commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1019/

All failures are on unrelated tests that were cancelled

https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1019/artifact/testlog/x86_64_release/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1018/

kbr- commented:
> All failures are on unrelated tests that were cancelled
> https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1019/artifact/testlog/x86_64_release/

May be timeouts. See in the longs how long it took from these tests starting to the cancel.
I'm apparently getting the same thing on https://github.com/scylladb/scylla/pull/10861#issuecomment-1165532741 right now.

I suspect it's because we overload the jenkins machine - by adding this new pool of nodes for running topology tests, concurrently with the existing tests.

alecco commented:
I pushed a rebase on latest master including Kamil's "no such keyspace" fix

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1089/
